Implementation: Basic quic-go in trxer

monfron opened the pull request:
Extended trxer with the quic-go lib. We can do the following:
- Quic connection setup and teardown
- Create quic streams
- Send data between Server and Client
- Measure throughput   

What we cannot do:
- Match the linerate (bad with loopback, worse with direct ethernet connection) 
- The best performance is achieved loopback on the i7 with 8 x 3,4 GHZ (240 MByte/s) 
- When I benchmark with the direct connection the "bad" workstations might be a bottleneck. 
- see additional information in commit msg

monfron committed:
Added: Basic function layout of quic client/server side and api parsing

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: quic client framework

- framework for creating several quic streams
- quic-go not "mounted" atm
- solved "problem" with goroutines and sync primitive

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implemented: Quic Client side based on quic-go

Still....verify if session.Close() needed!
Clemente et. al dont use it!
Verify it with server counterpart

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Debug code and unnecessary comments

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Fixed: Goroutine blocking issue

Handing over now the waitgroup counter as a reference.
I.e. no copies of waitgroup counter.
Result: Every thread now has the same knowledge, the wg.Done()
now successfully decrements and wg.Wait() stops blocking
after zero is reached.

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: quic_server() func

This function is based on the trxer udp_server() func.
It is starting several go routines based on the desired ports.
I also played around with the go routines to get a feeling
how to interact with them

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Debug output

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added: generate tls config func

function which generates key and cert step by step
and combines it to a tls cert.

Note: Lot of debug code going on, use it to verify
whats going on. Will be removed in the next commit!

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Debug code from create_tls_config() func

check out last commit to play around and verify functionality!

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Added and Tested: Quic connection establishment and teardown

A quic connection on a single stream is established step by step.
Lot of debug code going on! check it out to verify functionality.
Will be removed in the next commit!

The connection establishment and teardown is checked!

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Implemented: quic_server_worker() receiving now data

After setting up the connection and tearing it down, data can now
be sent.

Still the linerate is not matched. With increasing threads the
throughput also increased: c.f.
1 thread => 55 MByte/sec
2 thread => 93 MByte/sec
3 thread => 121 MByte/sec
...
Cap after 5 threads:
5 threads => 140 MByte/sec

Lot of debug code going on will be removed in the next commit

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Debug code and comments

For testing and further explanations check out the previous commit

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Removed: Additional debug code and comments

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Merge pull request #1 from monfron/quic-implementation

Implementation: Basic quic-go in trxer

hgn commented on trxer.go:
```diff
@@ -18,7 +29,161 @@ type measurement struct {
+func quic_client_worker(addr string, wg *sync.WaitGroup) {
+	fmt.Println("Quic stream connecting to: ", addr)
+
+	buf := make([]byte, 1400, 1400)
```
Probably this is a performance problem! You loop and sent 1400 byte in one call. The UDP code uses `BYTE_BUFFER_SIZE` which is quite larger.

This code should not just send 1400 bytes, just use BYTE_BUFFER_SIZE.

Better: we define a new application option (e.g. --call-size <n>) and let the user define the actual size in bytes.

hgn commented on trxer.go:
```diff
@@ -18,7 +29,161 @@ type measurement struct {
+
+func quic_server_worker(c chan<- measurement, port int) {
+	var bytesPerInterval uint64 = 0
+	buf := make([]byte, 1400, 1400)
```
same here, this should be replaced by BYTE_BUFFER_SIZE.

Or better: make this dynamic again (like in the sending process).

One note: when dynamic argument, then the argument should also work for UDP and TCP.

hgn reviewed (changes requested):
The code looks fine, except this one little but important comment! Please check this and add an follow up patch.

Thank you!

Commit:
Fixed: UDP performance issue

- increased application buffer size to BYTE_BUFFER_SIZE
(46 times bigger datagram now written to kernel!)
- future commit implements additional param to change that dynamically

Signed-off-by: Simon Georg Fronhöfer <s.fronhoefer@campus.tu-berlin.de>

Commit:
Improved: QUIC performance issues

- Same thing applied to QUIC as for UDP (increase application buffer)
- The performance then doubles
- But still the throughput is bad
- In the next commit I will implement a param to change the
application buffer dynamically

Signed-off-by: Simon Georg Fronhöfer <s.fronhoefer@campus.tu-berlin.de>

Commit:
Added: Dynamic call-size/application buffer size param

The call-size can be chosen between 4096 and 212992 byte.
Note that the UDP api (Dial() and consequent write()) results in a panic
when the max limit is reached. This should not be the case
(I checked the values with the kernel)

Signed-off-by: Simon Georg Fronhöfer <s.fronhoefer@campus.tu-berlin.de>

hgn commented on trxer.go:
```diff
@@ -22,17 +22,20 @@ import "io"
 
-var BYTE_BUFFER_SIZE = 8096 * 8
+var DEF_BUFFER_SIZE = 8096 * 8
+var MIN_BUFFER_SIZE = 4096
```
why an *artifical* minimum? If the user want to test 1 byte transfer, why the hell not?

monfron replied:
with write() we copy all data from application buffer to socket buffer right? then: when i check out for example the tcp_wmem variable I see 4096 is the minimum TCP send buffer space. So when the write() syscall comes in and the send buffer is empty...we instantly fill it up completely. with just one byte its a lot of overhead isnt it?

hgn replied:
Nope, not really. It is not *that* true, but true enough to explain the stuff:

`tcp_wmem` sets a minimum buffer size, this is the kernel internal buffer size allocated. But it is *no* limit to the syscall. The user can always send 1 byte of application data to the socket via write. `tcp_wmem` is just an kernel internal buffer tuning knob. And think about twice: you set an artificical limit in your application because you stumbled over `tcp_wmem`. Normally application programmer doent even know about `tcp_wmem`. So please remove the min limit, also the upper limit should be removed. The upper limit you encountered is probably just a *actual* limit, and will be removed in the future. But your application limit will stay forever. Another questions that arise: what about Windows, *BSD, ..? There are probably other limits somewhere? Hardcodeing a _current linux_ limits makes no sense again.

Update: it makes sense when your application has **application limits**, but you should not encode probably operating system limits to your applications again. If the operating system has problems it will raise an error and this error should be presented to the user. So that the user can react and start the application again with an appropriate value.

hgn commented on trxer.go:
```diff
@@ -22,17 +22,20 @@ import "io"
-var BYTE_BUFFER_SIZE = 8096 * 8
+var DEF_BUFFER_SIZE = 8096 * 8
+var MIN_BUFFER_SIZE = 4096
+var MAX_BUFFER_SIZE = 212992
```
What? Where comes the limit in? A can send 4GB into the TCP stack write() system call, right?

hgn commented:
If the limits are removed I will merge the changes - thank you @monfron!

monfron committed:
Removed: Application buffer limits

I removed the application buffer limits, that are based on
the kernel variables (tcp_wmem, rmem_max)

Signed-off-by: monfron <simon.fronhoefer@gmail.com>

monfron committed:
Merge pull request #2 from monfron/quic-implementation

Removed: Application buffer limits

hgn commented:
Thank you, merged!
